### PR TITLE
1385 and 1386 button updates

### PR DESCRIPTION
**Pull request by toni-leigh**
## Description

d8ea3eb
* Uses display: flex for Button text vertical centring
* In preparation for #1385 and #1386 where the text size will become more flexible

e90261b
* Allows smaller font to be set in large Buttons

a019c34
* Allows two-tier button with subtext

Note: the one pixel difference in internal alignment is because I let flex do it's thing centring rather than adding an extra pixel of pushing around, which I didn't think was a good idea.

## TODO
- [x] Release notes

## Related Issues / Pull Requests
#1385 
#1386 

## Screenshots / Gifs

Left is the design, right is the demo site:
image

## Testing Instructions
* Use the demo site: check the `subtext` prop will add the text as expected, but only if the button size is large

### Browsers

- [ ] IE11
- [ ] Edge
- [ ] Safari
- [ ] Chrome
- [ ] Firefox

### QA Issues

- [x] flex problems in Edge
- [x] flex problems in Safari
- [x] `smallFont` prop being sent through to Link erroneously

### CODE REVIEW

File: src/components/button/button.js [hunk 1]
```diff
@@ -58,6 +58,21 @@ class Button extends React.Component {
     disabled: PropTypes.bool,
 
     /**
+     * Allows a font size to be set that alters the default font size.
+     * Currently only setting a smalleer font in a large button is allowed, which we do with CSS
```
Comment (adamgeorgeson) [inline src/components/button/button.js]: `smaller`

File: src/components/button/button.js [hunk 2]
```diff
@@ -73,7 +88,22 @@ class Button extends React.Component {
      * @type {String}
      * @default medium
      */
-    size: PropTypes.string
+    size: PropTypes.string,
+
+    /**
+     * Allows a font size to be set that alters the default font size.
+     * Currently only setting a smalleer font in a large button is allowed, which we do with CSS
```
Comment (adamgeorgeson) [inline src/components/button/button.js]:
```
Allows subtext to be set.
Currently only allowing when button is large.
```

File: src/components/button/button.js [hunk 3]
```diff
@@ -58,6 +58,21 @@ class Button extends React.Component {
     disabled: PropTypes.bool,
 
     /**
+     * Allows a font size to be set that alters the default font size.
+     * Currently only setting a smalleer font in a large button is allowed, which we do with CSS
+     *
+     * @property fontSize
```
Comment (abramin) [inline src/components/button/button.js]: property name doesn't match

File: src/components/button/button.js [hunk 4]
```diff
@@ -73,7 +88,22 @@ class Button extends React.Component {
...
      * @default medium
      */
-    size: PropTypes.string
+    size: PropTypes.string,
+
+    /**
+     * Allows a font size to be set that alters the default font size.
+     * Currently only setting a smalleer font in a large button is allowed, which we do with CSS
+     *
+     * @property fontSize
```
Comment (abramin) [inline src/components/button/button.js]: This documentation doesn't apply to `subtext`

File: src/components/button/button.js [hunk 5]
```diff
@@ -58,6 +58,21 @@ class Button extends React.Component {
     disabled: PropTypes.bool,
 
     /**
+     * Allows a font size to be set that alters the default font size.
+     * Currently only setting a smalleer font in a large button is allowed, which we do with CSS
+     *
+     * @property fontSize
+     * @type {String}
```
Comment (abramin) [inline src/components/button/button.js]: Should this be Boolean?

Comment (adamgeorgeson) [inline src/components/button/button.js]: I was wondering this too. The return values made me question myself.

File: CHANGELOG.md [hunk 6]
```diff
@@ -171,6 +171,8 @@ return (
 
 * `Alert` now alerts itself to screen readers.
 * `Browser`: add a new method `setInputFocus` to focus on the input field of passed in ref but does not select text
+* `Button`: Allows small text on a large button [#1386](https://github.com/Sage/carbon/issues/1386)
```
Comment (toni-leigh) [inline CHANGELOG.md]: Move these two lines to v1.3.0 in CHANGELOG.md

File: src/components/button/button.js [hunk 7]
```diff
@@ -73,7 +89,22 @@ class Button extends React.Component {
      * @type {String}
      * @default medium
      */
-    size: PropTypes.string
+    size: PropTypes.string,
+
+    /**
+     * Sets a second bit of text under the main text, feinter and smaller.
```
Comment (abramin) [inline src/components/button/button.js]: [minor] `fainter`

Comment (toni-leigh) [inline src/components/button/button.js]: They use feint to describe extremely lightly lined paper and I extrapolated from that, however, I might actually be wrong to use the word like this.

Comment (abramin) [inline src/components/button/button.js]: Really, I always thought `feint` meant something like a fake attack.

Comment (toni-leigh) [inline src/components/button/button.js]: https://www.collinsdictionary.com/dictionary/english/feint-ruled-paper

I'll change this though I think this is less internationally understandable

**Comment by toni-leigh** [comment]
@abramin the `smallFont` prop is required because when the button is used alongside one with split text - [the issue](https://github.com/Sage/carbon/issues/1386) has more detail

**Comment by toni-leigh** [comment]
They're actually two separate things, `smallFont` is not used to govern the font-size of the two tier button

**Comment by abramin** [comment]
I didn't entirely understand the issues. When you say split text do you mean `subText`?  
I see now that the `smallFont` prop is separate but I'm still not clear about its purpose? 
Do we always want to keep font size the same between large and medium buttons perhaps? Maybe worth checking with UX.

**Comment by toni-leigh** [comment]
@abramin It's separate because I want to be able to control the font-size of one button so that the font can be set to a size which is visually coherent when that button is used with a second button that has two tier text. This prop defaults to false, so the button text will stay large unless purposely chosen.

It's feasible someone could use this prop on it's own, which I didn't think was a terrible thing to offer the developer, it might come out in other designs too, these things often do.

File: src/components/button/button.scss [hunk 8]
```diff
@@ -35,9 +38,20 @@
   }
 
   &.carbon-button--large {
-    font-size: 16px;
+    font-size: $app-font-size;
```
Comment (abramin) [inline src/components/button/button.scss]: could we just not remove this line? As it's set in the parent class,

File: src/components/button/button.scss [hunk 9]
```diff
@@ -5,11 +5,14 @@
 @import "./../../style-config/buttons";
 
 .carbon-button {
+  align-items: center;
   border: $carbon-button__border--secondary;
   border-radius: $carbon-button__border-radius;
   box-sizing: border-box;
   cursor: pointer;
-  display: inline-block;
+  display: inline-flex;
```
Comment (abramin) [inline src/components/button/button.scss]: These CSS changes seem to have broken the button contents in Safari.
screen shot 2017-07-12 at 09 09 24
screen shot 2017-07-12 at 09 09 27

Comment (toni-leigh) [inline src/components/button/button.scss]: @abramin Thanks for the cross browser checking!

Could I have some more details on context and steps to reproduce both of the above?

Comment (abramin) [inline src/components/button/button.scss]: The screenshots are from the demo the site in Safari. To see the hamburger menu you need to shrink the browser a bit.

Comment (toni-leigh) [inline src/components/button/button.scss]: Right thanks for that, I'll get a look

Comment (abramin) [inline src/components/button/button.scss]: Looks fine in Chrome & Firefox.

Menu button broken in Edge but regular button looks fine.

Don't have IE11 to test locally.

Comment (toni-leigh) [inline src/components/button/button.scss]: I've updated the PR description so this can be more easily tracked

**Pull request merged**